### PR TITLE
Backport PITR to 6X

### PR DESCRIPTION
**Pull request by x4m**
Hi!
@nastprol is working on the support of PITR of GP in [WAL-G](https://github.com/wal-g/wal-g/pull/992). And that's cool.
She is using gp_create_restore_point() which is not present in v6. Can we backport this small function?
Here's PoC for this backport. I had to move the function to gpcontrib to avoid interference with the system catalog.

What do you think?

Many thanks!

### CODE REVIEW

**Comment by ashwinstar** [comment]
This seems great effort. I will review it tomorrow.

**Comment by ashwinstar** [comment]
> Hi!
> @nastprol is working on the support of PITR of GP in [WAL-G](https://github.com/wal-g/wal-g/pull/992). And that's cool.

I high-level can guess how this integration is being thought on. We would love to hear more explicitly on it as the pointed PR in WAL-G has no documentation currently.

**Comment by x4m** [comment]
We expect that the user will set up WAL archiving on each segment.
Greenplum ```wal-g backup-push`` will do:
1. Physical backup of each segment as if it's a PostgreSQL cluster to separate folder for each segment.
2. Call gp_create_restore_point().
3. Store sentinel file with meta information.

There will be no separate backup-fetch, each segment is restored as if it's a PostgreSQL node.

**Comment by maksm90** [comment]
> 2. Call gp_create_restore_point().

Do you intend to achieve PITR via periodical further calls of `gp_create_restore_point()`?

**Comment by x4m** [comment]
Probably, but we are not going to automate this in WAL-G...

File: src/include/catalog/pg_proc_gp.h [hunk 1]
```diff
@@ -983,9 +983,5 @@ DESCR("Legacy cdbhash function");
 DATA(insert OID = 6171 ( cdblegacyhash_anyenum  PGNSP PGUID 12 1 0 0 0 f f f f t f i 1 0 23 "3500" _null_ _null_ _null_ _null_ cdblegacyhash_anyenum _null_ _null_ _null_ n a ));
 DESCR("Legacy cdbhash function");
 
-/* gp_create_restore_point(text) => SETOF record */
```
Comment (asimrp) [inline src/include/catalog/pg_proc_gp.h]: Ooops, this is a catalog version bump.

Comment (asimrp) [reply src/include/catalog/pg_proc_gp.h]: Somehow, I ended up reviewing outdated version of the patch, please ignore the above comment.

**Comment by iJKos** [comment]
Looked quickly, it seems to be what we need so much. I can't wait for this to be released :)

**Comment by x4m** [comment]
@ashwinstar is there any news on this PR? Is general design OK or should we change something? WAL-G part is nearing completion, btw. Currently, we support only 7+, but 6 would be awesome.

**Review by jimmyyih — APPROVED** [review]
I think this should be fine to merge in.  It works as expected.

@x4m Could you resolve the trivial spelling comment and squash the 4 commits (25aad97, 5423db7, d529265, and d1401b9) into 1 commit?  In the meantime, we'll put this through our merge approval process (required for stable branches but not for master).

**Comment by jimmyyih** [comment]
> We expect that the user will set up WAL archiving on each segment.

Curious... does this imply using `archive_command` or just setting up the segments for `pg_basebackup` (e.g. replication user in `pg_hba.conf`)?  If you're implying `archive_command`, there may be a lot of unknowns on how that will work for GPDB 6X.  For GPDB 7X, we are currently doing a lot of work to get `archive_command` and `restore_command` to work seamlessly in a distributed way (e.g. https://github.com/greenplum-db/gpdb/commit/96b9844bc803162edded0fe39755e2f97178c45d)... but there are no current plans on having that work backported to 6X.

**Comment by x4m** [comment]
> I think this should be fine to merge in. It works as expected.
Thanks for the reiview!


> @x4m Could you resolve the trivial spelling comment and squash the 4 commits ([25aad97](https://github.com/greenplum-db/gpdb/commit/25aad97049f135aac1f5bf7880c62281a543985e), [5423db7](https://github.com/greenplum-db/gpdb/commit/5423db713b4aca7f44cd35a43a60804950d8f9e4), [d529265](https://github.com/greenplum-db/gpdb/commit/d529265dcf9be8c0b52a777ab68556c387824db1), and [d1401b9](https://github.com/greenplum-db/gpdb/commit/d1401b9ec61a99678f36d7ca91b09f60918a7762)) into 1 commit? In the meantime, we'll put this through our merge approval process (required for stable branches but not for master).
Here's b881526




> > We expect that the user will set up WAL archiving on each segment.
> 
> Curious... does this imply using `archive_command` or just setting up the segments for `pg_basebackup` (e.g. replication user in `pg_hba.conf`)? If you're implying `archive_command`, there may be a lot of unknowns on how that will work for GPDB 6X. For GPDB 7X, we are currently doing a lot of work to get `archive_command` and `restore_command` to work seamlessly in a distributed way (e.g. [96b9844](https://github.com/greenplum-db/gpdb/commit/96b9844bc803162edded0fe39755e2f97178c45d))... but there are no current plans on having that work backported to 6X.

I think we are OK if archive_command will work as it does in Postres. ```contentId``` seems like a useful invention, but we can emulate this via WAL-G config. Our installation runs on 6.15, so we have to deal with it in 6 :)

Anyway, we will have WAL-G in charge of deleting WALs and backups from S3. And this implies that we have to set up different config for each segment.

File: gpcontrib/gp_pitr/gp_pitr.control [hunk 2]
```diff
@@ -0,0 +1,3 @@
+comment = 'Backport of PITR capabilities from Greenplum 7.0'
```
Comment (kmacoskey) [inline gpcontrib/gp_pitr/gp_pitr.control:1]: Not really reviewing. Just wanted to point out that the "Comment" here should be a description of what the extension does. It shouldn't be a description of why it is being added.

Comment (x4m) [reply gpcontrib/gp_pitr/gp_pitr.control]: Thanks! I've updated comment, plz let me know if you have better wording, English is not my native language.

**Review by shivzone — APPROVED** [review]
Looks good

**Comment by usernamedt** [comment]
I've tried performing backup with PITR using the `pg_start_backup()/pg_stop_backup()` API on 6X and found that there is one line missing from the `pg_stop_backup()` handler function. Possibly it was lost in [this PG9.4 merge commit](https://github.com/greenplum-db/gpdb/commit/1617960e8ab077988c26bc8d59c5cfbfa608a3fb#diff-5bf8f0a43cee95c7fe2454e46f3a896aaf5e5b3158a759a1d113251a98396d16). Lack of this `do_pg_stop_backup()` call causes the pg_stop_backup malfunctioning so user is never able to end the started backup.

So I just restored the appropriate call from [Postgres 9.4 source code](https://github.com/postgres/postgres/blob/REL9_4_STABLE/src/backend/access/transam/xlogfuncs.c#L90). Now it works as expected. 

I attached the fix to this PR because the end goal is the same: backport PITR functionality to the 6X.

**Comment by jimmyyih** [comment]
Sorry for the wait; there was some small concern about how the new TwophaseCommitLock would affect performance on our stable branch.  Doing some pgbench concurrency testing with and without the patch, there wasn't really much of a noticeable difference in TPS... which is expected since the distributed commit prepared step is usually pretty fast and light weight.

With that said, this PR is good to merge with the expectation that the extension is primarily for WAL-G and will probably be maintained as such.  Thanks for the contribution and looking forward to more PRs relating to WAL-G.

**Comment by x4m** [comment]
Coo! Many thanks!

**Pull request merged**